README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# npboottprmFBar

<!-- badges: start -->
<!-- badges: end -->

The goal of ‘npboottprmFBar’ is to implement the nonparametric bootstrap
test with pooled resampling method (as presented in Dwivedi,
Mallawaarachchi, and Alvarado (2017)) for informative hypothesis testing
(as implemented in ‘restriktor’ and outlined in Vanbrabant and Rosseel
(2020)).

## Installation

You can install the released version of ‘npboottprmFBar’ from
[CRAN](https://CRAN.R-project.org):

``` r
install.packages("npboottprmFBar")
```

To install the development version of ‘npboottprmFBar’ from GitHub, use
the [devtools](https://devtools.r-lib.org/) package:

``` r
# install.packages("devtools")
devtools::install_github("mightymetrika/npboottprmFBar")
```

## An iris example

The following example demonstrates how to use the bootFbar() function to
conduct an informative hypothesis test.

``` r
library(npboottprmFBar)

res <- bootFbar(data = iris, formula = Sepal.Length ~ -1 + Species,
                grp = "Species",
                constraints = 'Speciessetosa < Speciesversicolor < Speciesvirginica',
                nboot = 10, conf.level = 0.95, seed = NULL, na_rm = FALSE)

paste0("Type B Test: ", res$pvalueB)
#> [1] "Type B Test: 1"
paste0("Type A Test: ", res$pvalueA)
#> [1] "Type A Test: 0"
```

The non-significant Type B test followed by the significant Type A test
is evidence in favor the order-constrained hypothesis

## References

Dwivedi AK, Mallawaarachchi I, Alvarado LA (2017). “Analysis of small
sample size studies using nonparametric bootstrap test with pooled
resampling method.” Statistics in Medicine, 36 (14), 2187-2205.

Vanbrabant, L., & Rosseel, Y. (2020). An Introduction to Restriktor:
Evaluating informative hypotheses for linear models. In R. van de Schoot
& M. Miocevic (Eds.), Small Sample Size Solutions: A Guide for Applied
Researchers and Practitioners (1st ed., pp. 157 -172). Routledge.
<https://doi.org/10.4324/9780429273872-14>


NEWS.md
# npboottprmFBar (development version)

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(npboottprmFBar)

test_check("npboottprmFBar")


tests/testthat/test-bootFbar.R
test_that("bootFbar works as expected with inequality constraints", {

  res <- bootFbar(data = iris, formula = Sepal.Length ~ -1 + Species,
                  grp = "Species",
                  constraints = 'Speciessetosa < Speciesversicolor < Speciesvirginica',
                  nboot = 10, conf.level = 0.95, seed = NULL, na_rm = FALSE)

  expect_equal(length(res), 11)
  expect_null(res$TsF)
  expect_true(is.na(res$pvalueF))

})

test_that("bootFbar works as expected when constraint only has equalities", {

  res <- bootFbar(data = iris, formula = Sepal.Length ~ -1 + Species,
                  grp = "Species",
                  constraints = 'Speciessetosa == Speciesversicolor == Speciesvirginica',
                  nboot = 10, conf.level = 0.95, seed = NULL, na_rm = FALSE)

  expect_equal(length(res), 11)
  expect_null(res$TsA)
  expect_null(res$TsB)
  expect_true(is.na(res$pvalueA))
  expect_true(is.na(res$pvalueB))

})

test_that("bootFbar's na_rm functions as expected", {

  # Get regular model with default iris & na_rm
  res <- bootFbar(data = iris, formula = Sepal.Length ~ -1 + Species,
                  grp = "Species",
                  constraints = 'Speciessetosa == Speciesversicolor == Speciesvirginica',
                  nboot = 10, conf.level = 0.95, seed = NULL, na_rm = FALSE)

  # Create missing data on var included in formula
  iris_miss <- iris
  iris_miss[1:5, "Sepal.Length"] <- NA

  # Run on iris_miss with na_rm = TRUE
  res_miss <- bootFbar(data = iris_miss, formula = Sepal.Length ~ -1 + Species,
                  grp = "Species",
                  constraints = 'Speciessetosa == Speciesversicolor == Speciesvirginica',
                  nboot = 10, conf.level = 0.95, seed = NULL, na_rm = TRUE)

  expect_equal(nrow(res$modelo$model) - 5, nrow(res_miss$modelo$model))

  rm(iris_miss)

  # Create missing data on var excluded from formula
  iris_miss <- iris
  iris_miss[1:5, "Petal.Length"] <- NA

  # Run on iris_miss with na_rm = TRUE
  res_miss <- bootFbar(data = iris_miss, formula = Sepal.Length ~ -1 + Species,
                       grp = "Species",
                       constraints = 'Speciessetosa == Speciesversicolor == Speciesvirginica',
                       nboot = 10, conf.level = 0.95, seed = NULL, na_rm = TRUE)

  expect_equal(nrow(res$modelo$model), nrow(res_miss$modelo$model))

  rm(iris_miss)


})


tests/testthat/test-persimon.R
test_that("persimon works", {

  set.seed(135)
  res <- persimon(M1 = 5, S1 = 1, M2 = 5, S2 = 1, M3 = 5, S3 = 1,
                  n1 = c(6,7), n2 = c(6,7), n3 = c(6,7),
                  n_simulations = 2, nboot = 5, conf.level = 0.95)


  expect_equal(nrow(res), 2)
  expect_equal(ncol(res), 16)
})

test_that("persimon works with non-null skew", {

  set.seed(1265)
  res <- persimon(M1 = 5, S1 = 1, M2 = 5, S2 = 1, M3 = 5, S3 = 1,
                  Sk1 = 0.8, Sk2 = 0.8, Sk3 = 0.8,
                  n1 = c(6,7), n2 = c(6,7), n3 = c(6,7),
                  n_simulations = 2, nboot = 5, conf.level = 0.95)


  expect_equal(nrow(res), 2)
  expect_equal(ncol(res), 16)
})


R/persimon.R
#' Performance Simulation On Type I Error or Statistical Power
#'
#' Simulates performance (Type I Error or Statistical Power) for a range of
#' statistical tests, including bootFbar, default setting restriktor::iht, and
#' parametric bootstrap restriktor::iht. The function is set up to replicate the
#' table structure in Dwivedi et al. (2017) Supplemental Tables 2 & 3.
#'
#' @param M1 Mean value for group 1.
#' @param M2 Mean value for group 2.
#' @param M3 Mean value for group 3.
#' @param S1 Standard deviation for group 1.
#' @param S2 Standard deviation for group 2.
#' @param S3 Standard deviation for group 3.
#' @param Sk1 Skewness parameter for group 1; NULL for normal distribution.
#' @param Sk2 Skewness parameter for group 2; NULL for normal distribution.
#' @param Sk3 Skewness parameter for group 3; NULL for normal distribution.
#' @param n1 Vector of sample sizes for group 1.
#' @param n2 Vector of sample sizes for group 2.
#' @param n3 Vector of sample sizes for group 3.
#' @param n_simulations Number of simulations to run (default is 10000).
#' @param nboot Number of bootstrap samples (default is 1000).
#' @param conf.level Confidence level for the tests (default is 0.95).
#'
#' @return A data frame with the proportions of rejecting the null hypothesis for
#' each test and sample size combination.
#'
#' @examples
#' set.seed(135)
#' persimon(M1 = 5, S1 = 1, M2 = 5, S2 = 1, M3 = 5, S3 = 1,
#'          n1 = 6, n2 = 6, n3 = 6, n_simulations = 2, nboot = 4,
#'          conf.level = 0.95)
#'
#' @details
#' The `persimon` function generates data for three groups with specified mean, standard deviation,
#' and skewness, and then applies a range of statistical tests to this data, simulating the process
#' across a specified number of iterations to assess performance in terms of Type I Error or
#' Statistical Power. The output mirrors the format of the supplemental tables in Dwivedi et al. (2017).
#'
#' @references
#' Dwivedi, A. K., Mallawaarachchi, I., & Alvarado, L. A. (2017). Analysis of small sample size studies
#' using nonparametric bootstrap test with pooled resampling method. Statistics in Medicine, 36(14), 2187–2205.
#' https://doi.org/10.1002/sim.7263
#'
#' @importFrom lmPerm lmp
#'
#' @export
persimon <- function(M1 = 5, S1 = 1, M2 = 5, S2 = 1, M3 = 5, S3 = 1,
                     Sk1 = NULL, Sk2 = NULL, Sk3 = NULL,
                     n1 = c(2,3,4,5,6,7,8,9,10,15),
                     n2 = c(2,3,4,5,6,7,8,9,10,15),
                     n3 = c(2,3,4,5,6,7,8,9,10,15),
                     n_simulations = 10000, nboot = 1000,
                     conf.level = 0.95){

  # Ensure n1 and n2 and n3 are of equal length
  if (!(length(n1) == length(n2) & length(n2) == length(n3))) {
    stop("n1 and n2 and n3 must be of equal length")
  }

  # Get results for one iteration of the simulation or one pair of sample sizes
  get_result <- function(n1, n2, n3) {

    # Generate normal or skew normal data
    generate_data <- function(n, mean, sd, skew) {
      if (is.null(skew)) {
        stats::rnorm(n, mean, sd)
      } else {
        fGarch::rsnorm(n, mean, sd, xi = skew)
      }
    }

    F1 <- generate_data(n1, M1, S1, Sk1)
    F2 <- generate_data(n2, M2, S2, Sk2)
    F3 <- generate_data(n3, M3, S3, Sk3)

    df <- data.frame(x = c(F1, F2, F3), grp = rep(c("F1", "F2", "F3"),
                                                  c(n1, n2, n3)))

    # Test p-value for each method
    npbft <- tryCatch({
      npboottprm::nonparboot(df, x = "x", grp = "grp", nboot = nboot, test = "F",
                             conf.level = conf.level)$p.value <= 1 - conf.level
    }, error = function(e) NA)

    anov <- tryCatch({
      stats::anova(stats::lm(x ~ factor(grp), df))$`F value`[1] <= 1 - conf.level
    }, error = function(e) NA)

    kw <- tryCatch({
      stats::kruskal.test(x ~ factor(grp), df)$p.value <= 1 - conf.level
    }, error = function(e) NA)

    pft <- tryCatch({
      lmPerm::aovp(x ~ factor(grp),df)$perm$P[2,1] <= 1 - conf.level
    }, error = function(e) NA)

    bFbar0 <- tryCatch({
      bootFbar(df, formula = "x ~ -1 + grp", grp = "grp",
               constraints = 'grpF1 == grpF2 == grpF3',
               nboot = nboot, conf.level = conf.level)$pvalueF <= 1 - conf.level

    }, error = function(e) NA)

    bFbar1 <- tryCatch({
      bout1 <- bootFbar(df, formula = "x ~ -1 + grp", grp = "grp",
                        constraints = 'grpF1 < grpF3', nboot = nboot,
                        conf.level = conf.level)

      bout1$pvalueB > 1 - conf.level & bout1$pvalueA <= 1 - conf.level

    }, error = function(e) NA)

    bFbar2 <- tryCatch({
      bout2 <- bootFbar(df, formula = "x ~ -1 + grp", grp = "grp",
                        constraints = 'grpF1 < grpF2 < grpF3',
                        nboot = nboot, conf.level = conf.level)

      bout2$pvalueB > 1 - conf.level & bout2$pvalueA <= 1 - conf.level

    }, error = function(e) NA)

    rFbar0 <- tryCatch({
      restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                      constraints = 'grpF1 == grpF2 == grpF3')$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    rFbar1 <- tryCatch({
      rout1 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                               constraints = 'grpF1 < grpF3')

      rout1$B$pvalue > 1 - conf.level & rout1$A$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    rFbar2 <- tryCatch({
      rout2 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                               constraints = 'grpF1 < grpF2 < grpF3')

      rout2$B$pvalue > 1 - conf.level & rout2$A$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    rbFbar0 <- tryCatch({
      restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                      constraints = 'grpF1 == grpF2 == grpF3',
                      boot = "parametric", R = nboot)$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    rbFbar1 <- tryCatch({
      rbout1 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                                constraints = 'grpF1 < grpF3',
                                boot = "parametric", R = nboot)

      rbout1$B$pvalue > 1 - conf.level & rbout1$A$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    rbFbar2 <- tryCatch({
      rbout2 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                                constraints = 'grpF1 < grpF2 < grpF3',
                                boot = "parametric", R = nboot)

      rbout2$B$pvalue > 1 - conf.level & rbout2$A$pvalue <= 1 - conf.level

    }, error = function(e) NA)

    # Returning a named vector
    return(c(NPBFT = npbft, ANOV = anov, KW = kw, PFT = pft,
             bFBAR0 = bFbar0, bFBAR1 = bFbar1, bFBAR2 = bFbar2,
             rFBAR0 = rFbar0, rFBAR1 = rFbar1, rFBAR2 = rFbar2,
             rbFBAR0 = rbFbar0, rbFBAR1 = rbFbar1, rbFBAR2 = rbFbar2))
  }

  # Run n_simulations at each sample size pair
  results <- lapply(seq_along(n1), function(i) {
    replicate(n_simulations, get_result(n1[i], n2[i], n3[i]))
  })

  # Use lapply to calculate proportions and create a data frame for each sample size combination
  proportions_list <- lapply(seq_along(results), function(i) {
    proportions <- rowMeans(results[[i]], na.rm = TRUE)
    data.frame(
      n1 = n1[i],
      n2 = n2[i],
      n3 = n3[i],
      ANOV = proportions["ANOV"],
      KW = proportions["KW"],
      NPBFT = proportions["NPBFT"],
      PFT = proportions["PFT"],
      bFBAR0 = proportions["bFBAR0"],
      bFBAR1 = proportions["bFBAR1"],
      bFBAR2 = proportions["bFBAR2"],
      rFBAR0 = proportions["rFBAR0"],
      rFBAR1 = proportions["rFBAR1"],
      rFBAR2 = proportions["rFBAR2"],
      rbFBAR0 = proportions["rbFBAR0"],
      rbFBAR1 = proportions["rbFBAR1"],
      rbFBAR2 = proportions["rbFBAR2"]
    )
  })

  # Combine the list of data frames into a single data frame
  proportions_df <- do.call(rbind, proportions_list)

  return(proportions_df)
}


R/persimon_TypeI.R
persimon_TypeI <- function(M1 = 5, S1 = 1, M2 = 5, S2 = 1, M3 = 5, S3 = 1,
                             Sk1 = NULL, Sk2 = NULL, Sk3 = NULL,
                             n1 = c(2,3,4,5,6,7,8,9,10,15),
                             n2 = c(2,3,4,5,6,7,8,9,10,15),
                             n3 = c(2,3,4,5,6,7,8,9,10,15),
                             n_simulations = 10000, nboot = 1000,
                             conf.level = 0.95){

  # Ensure n1 and n2 and n3 are of equal length
  if (!(length(n1) == length(n2) & length(n2) == length(n3))) {
    stop("n1 and n2 and n3 must be of equal length")
  }

  # Get results for one iteration of the simulation or one pair of sample sizes
  get_result <- function(n1, n2, n3) {

    # Generate normal or skew normal data
    generate_data <- function(n, mean, sd, skew) {
      if (is.null(skew)) {
        stats::rnorm(n, mean, sd)
      } else {
        fGarch::rsnorm(n, mean, sd, xi = skew)
      }
    }

    F1 <- generate_data(n1, M1, S1, Sk1)
    F2 <- generate_data(n2, M2, S2, Sk2)
    F3 <- generate_data(n3, M3, S3, Sk3)

    df <- data.frame(x = c(F1, F2, F3), grp = rep(c("F1", "F2", "F3"), c(n1, n2, n3)))

    # Test p-value for each method
    npbft <- tryCatch({
      npboottprm::nonparboot(df, x = "x", grp = "grp", nboot = nboot, test = "F", conf.level = conf.level)$p.value <= 1 - conf.level
    }, error = function(e) NA)

    anov <- tryCatch({
      stats::anova(stats::lm(x ~ factor(grp), df))$`F value`[1] <= 1 - conf.level
    }, error = function(e) NA)

    kw <- tryCatch({
      stats::kruskal.test(x ~ factor(grp), df)$p.value <= 1 - conf.level
    }, error = function(e) NA)

    pft <- tryCatch({
      lmPerm::aovp(x ~ factor(grp),df)$perm$P[2,1] <= 1 - conf.level
    }, error = function(e) NA)

    bFbar0 <- tryCatch({
      bootFbar(df, formula = "x ~ -1 + grp", grp = "grp", constraints = 'grpF1 = grpF2 = grpF3',
               nboot = nboot, conf.level = conf.level)$pvalueF <= 1 - conf.level

    }, error = function(e) NA)

    bFbar1 <- tryCatch({
      bout1 <- bootFbar(df, formula = "x ~ -1 + grp", grp = "grp", constraints = 'grpF1 < grpF3',
               nboot = nboot, conf.level = conf.level)
      bout1$pvalueB >= 1 - conf.level & bout1$pvalueA <= 1 - conf.level

    }, error = function(e) NA)

    bFbar2 <- tryCatch({
      bout2 <- bootFbar(df, formula = "x ~ -1 + grp", grp = "grp", constraints = 'grpF1 < grpF2 < grpF3',
               nboot = nboot, conf.level = conf.level)
      bout2$pvalueB >= 1 - conf.level & bout1$pvalueA <= 1 - conf.level

    }, error = function(e) NA)

    Fbar0 <- tryCatch({
      restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                      constraints = 'grpF1 = grpF2 = grpF3')$pvalue <= 1 - conf.level
    }, error = function(e) NA)

    Fbar1 <- tryCatch({
      out1 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                      constraints = 'grpF1 < grpF3')

      out1$B$pvalue >= 1 - conf.level & out1$A$pvalue <= 1 - conf.level
    }, error = function(e) NA)

    Fbar2 <- tryCatch({
      out2 <- restriktor::iht(stats::lm(formula = x ~ -1 + grp, data =df),
                              constraints = 'grpF1 < grpF2 < grpF3')

      out2$B$pvalue >= 1 - conf.level & out2$A$pvalue <= 1 - conf.level
    }, error = function(e) NA)

    # Returning a named vector
    return(c(NPBFT = npbft, ANOV = anov, KW = kw, PFT = pft,
             bFBAR0 = bFbar0, bFBAR1 = bFbar1, bFBAR2 = bFbar2,
             FBAR0 = Fbar0, FBAR1 = Fbar1, FBAR2 = Fbar2))
  }

  # Run n_simulations at each sample size pair
  results <- lapply(seq_along(n1), function(i) {
    replicate(n_simulations, get_result(n1[i], n2[i], n3[i]))
  })

  # Use lapply to calculate proportions and create a data frame for each sample size combination
  proportions_list <- lapply(seq_along(results), function(i) {
    proportions <- rowMeans(results[[i]], na.rm = TRUE)
    data.frame(
      n1 = n1[i],
      n2 = n2[i],
      n3 = n3[i],
      ANOV = proportions["ANOV"],
      KW = proportions["KW"],
      NPBFT = proportions["NPBFT"],
      PFT = proportions["PFT"],
      bFBAR0 = proportions["bFBAR0"],
      bFBAR1 = proportions["bFBAR1"],
      bFBAR2 = proportions["bFBAR2"],
      FBAR0 = proportions["FBAR0"],
      FBAR1 = proportions["FBAR1"],
      FBAR2 = proportions["FBAR2"]
    )
  })

  # Combine the list of data frames into a single data frame
  proportions_df <- do.call(rbind, proportions_list)

  return(proportions_df)
}


R/bootFbar.R
#' Nonparametric Bootstrap Test with Pooled Resampling for Informative Hypothesis Testing
#'
#' Performs a nonparametric bootstrap test with pooled resampling based on the methods
#' described by Dwivedi et al. (2017) and using the `restriktor::iht` function as
#' proposed by Vanbrabant and Rosseel (2020).
#'
#' @param data A data frame containing the dataset for analysis.
#' @param formula An object of class \code{"formula"} (or one that can be coerced to that class):
#' a symbolic description of the model to be fitted.
#' @param grp A string specifying the grouping variable in the data.
#' @param constraints A matrix or data frame of constraints for the hypothesis test.
#' @param nboot An integer indicating the number of bootstrap resamples (default is 1000).
#' @param conf.level A numeric value specifying the confidence level for the interval
#' (default is 0.95).
#' @param seed An optional integer setting the seed for random number generation (default is NULL).
#' @param na_rm A logical value indicating whether NA values should be removed (default is FALSE).
#'
#' @return A list containing the following components:
#'
#'   -pvalueA, pvalueB, pvalueF: P-values for the different test types.
#'
#'   -ciA, ciB, ciF: Confidence intervals for the test statistics.
#'
#'   -TsA, TsB, TsF: Bootstrap test statistics.
#'
#'   -modelo: The linear model object.
#'
#'   -ihto: The initial iht computation result.
#'
#' @examples
#' bootFbar(data = iris, formula = Sepal.Length ~ -1 + Species,
#'          grp = "Species",
#'          constraints = 'Speciessetosa < Speciesversicolor < Speciesvirginica',
#'          nboot = 10, conf.level = 0.95, seed = NULL, na_rm = FALSE)
#'
#' @references
#' Dwivedi, A. K., Mallawaarachchi, I., & Alvarado, L. A. (2017). Analysis of
#' small sample size studies using nonparametric bootstrap test with pooled
#' resampling method. Statistics in Medicine, 36(14), 2187–2205.
#' https://doi.org/10.1002/sim.7263
#'
#' Vanbrabant, L., & Rosseel, Y. (2020). An Introduction to Restriktor: Evaluating
#' informative hypotheses for linear models. In R. van de Schoot & M. Miocevic
#' (Eds.), Small Sample Size Solutions: A Guide for Applied Researchers and
#' Practitioners (1st ed., pp. 157 -172). Routledge. https://doi.org/10.4324/9780429273872-14
#'
#' @export
bootFbar <- function (data, formula, grp, constraints, nboot = 1000, conf.level = 0.95,
                      seed = NULL, na_rm = FALSE) {
  # Check parameters
  stopifnot("data must be a data frame" = is.data.frame(data))
  stopifnot("nboot must be a numeric integer greater than 0" = is.numeric(nboot) && length(nboot) == 1L && nboot > 0 && nboot == as.integer(nboot))
  stopifnot("conf.level must be a numeric value between 0 and 1" = is.numeric(conf.level) && length(conf.level) == 1L && conf.level > 0 && conf.level < 1)
  stopifnot("seed must be NULL or numeric" = is.null(seed) || (is.numeric(seed) && length(seed) == 1L))

  # Set seed
  if (!is.null(seed)) {
    set.seed(seed)
  }

  # Extract variable names from the formula
  vars_in_formula <- all.vars(stats::as.formula(formula))

  # Remove NA values if na_rm is TRUE, but only for the variables in the formula
  if (na_rm == TRUE) {
    subset_data <- data[, vars_in_formula, drop = FALSE]
    complete_cases <- stats::complete.cases(subset_data)
    data <- data[complete_cases, ]
  }

  grp_val <- data[[grp]]

  grp_sizes <- table(grp_val)

  # Initialize variables
  pvalueA <- pvalueB <- pvalueF <- NA
  ciA <- ciB <- ciF <- NULL
  bootResults_A <- bootResults_B <- bootResults_F <- NULL

  # Prepare the model and initial iht computation
  model <- stats::lm(stats::as.formula(formula), data = data)
  iht_orig <- restriktor::iht(model, constraints = constraints)

  # Check if only equality constraints are used
  only_equality_constraints <- is.null(iht_orig$B)

  # Pre-calculate original test statistics
  pre_calc <- list(orig_Ts_F = if (only_equality_constraints) iht_orig$Ts else NULL,
                   orig_Ts_B = if (!only_equality_constraints) iht_orig$B$Ts else NULL,
                   orig_Ts_A = if (!only_equality_constraints) iht_orig$A$Ts else NULL)

  # Perform bootstrap resampling
  bootstrap_results <- replicate(nboot, bootstrap_fbar_sample(data, formula, grp, constraints, grp_val, grp_sizes, pre_calc, only_equality_constraints))

  # Process results based on constraint type
  if (only_equality_constraints) {
    bootResults_F <- unlist(bootstrap_results)

    # Calculate the p-value and confidence interval for F-test
    pvalueF <- mean(abs(bootResults_F) >= abs(pre_calc$orig_Ts_F), na.rm = TRUE)
    ciF <- stats::quantile(bootResults_F, c((1 - conf.level) / 2, 1 - (1 - conf.level) / 2))
  } else {
    bootResults_B <- unlist(bootstrap_results[1, ])
    bootResults_A <- unlist(bootstrap_results[2, ])

    # Calculate the p-values and confidence intervals for Type A and B
    pvalueB <- mean(abs(bootResults_B) >= abs(pre_calc$orig_Ts_B), na.rm = TRUE)
    pvalueA <- mean(abs(bootResults_A) >= abs(pre_calc$orig_Ts_A), na.rm = TRUE)
    ciB <- stats::quantile(bootResults_B, c((1 - conf.level) / 2, 1 - (1 - conf.level) / 2))
    ciA <- stats::quantile(bootResults_A, c((1 - conf.level) / 2, 1 - (1 - conf.level) / 2))
  }

  return(list(pvalueA = pvalueA, pvalueB = pvalueB, pvalueF = pvalueF,
              ciA = ciA, ciB = ciB, ciF = ciF,
              TsA = bootResults_A, TsB = bootResults_B, TsF = bootResults_F,
              modelo = model, ihto = iht_orig))
}

#' Internal Bootstrap Resampling for F-bar Test
#'
#' This internal function performs bootstrap resampling specific to the F-bar test
#' within the context of the nonparametric bootstrap test with pooled resampling.
#' It is used by `bootFbar` to generate bootstrap samples and compute intermediate
#' statistics.
#'
#' @param data A data frame containing the dataset for analysis.
#' @param formula An object of class \code{"formula"}: a symbolic description of
#' the model to be fitted.
#' @param grp A string specifying the grouping variable in the data.
#' @param constraints A matrix or data frame of constraints for the hypothesis test.
#' @param grp_val A vector containing the values of the group variable.
#' @param grp_sizes A table object containing the sizes of each group.
#' @param pre_calc A list containing pre-calculated original test statistics.
#' @param only_equality_constraints A logical indicating whether only equality
#' constraints are used.
#'
#' @return A list containing the test statistics for each constraint type
#' (B, A, F) based on the bootstrap sample. Returns NA values if the iht
#' computation fails.
#'
#' @details
#' The function creates bootstrap samples for each group in the data, then binds
#' these samples together. It applies the `restriktor::iht` function on these
#' samples to compute the test statistics. Error handling is implemented to manage
#' potential issues in `iht` computation.
#'
#' @keywords internal
bootstrap_fbar_sample <- function(data, formula, grp, constraints, grp_val, grp_sizes, pre_calc, only_equality_constraints) {
  grp_unique <- unique(grp_val)

  # Create bootstrap samples for each group
  bootstrap_samples <- lapply(grp_unique, function(g) {
    data[sample(nrow(data), size = grp_sizes[g], replace = TRUE), ]
  })

  # Combine bootstrap samples and assign new groups
  bind_boot <- do.call(rbind, bootstrap_samples)
  new_group <- rep(grp_unique, times = sapply(bootstrap_samples, nrow))
  bind_boot[[grp]] <- new_group

  # Run model and use tryCatch to handle potential errors in iht_boot
  iht_boot <- tryCatch({
    restriktor::iht(stats::lm(stats::as.formula(formula), data = bind_boot), constraints = constraints)
  }, error = function(e) {
    return(NULL)
  })

  # Return result
  if (is.null(iht_boot)) {
    if (only_equality_constraints) {
      return(list(bsamp_Ts_F = NA))
    } else {
      return(list(bsamp_Ts_B = NA,
                  bsamp_Ts_A = NA))
    }
  }

  if (only_equality_constraints) {
    return(list(bsamp_Ts_F = iht_boot$Ts))
  } else {
    return(list(bsamp_Ts_B = iht_boot$B$Ts,
                bsamp_Ts_A = iht_boot$A$Ts))
  }
}
